File: src/main/java/net/onelitefeather/titan/function/SitFunction.java
package net.onelitefeather.titan.function;

import net.minestom.server.coordinate.Point;
import net.minestom.server.coordinate.Pos;
import net.minestom.server.entity.Entity;
import net.minestom.server.entity.EntityType;
import net.minestom.server.entity.Player;
import net.minestom.server.event.Event;
import net.minestom.server.event.EventDispatcher;
import net.minestom.server.event.EventNode;
import net.minestom.server.event.player.PlayerBlockInteractEvent;
import net.minestom.server.event.player.PlayerDisconnectEvent;
import net.minestom.server.event.player.PlayerPacketEvent;
import net.minestom.server.instance.block.Block;
import net.minestom.server.network.packet.client.play.ClientSteerVehiclePacket;
import net.onelitefeather.titan.event.EntityDismountEvent;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public final class SitFunction {

    private static final double SIT_OFFSET = 0.25;

    private final Map<UUID, Pos> playerExitLocations = new HashMap<>();
    private final Map<UUID, Entity> playerSitsMap = new HashMap<>();

    private SitFunction(EventNode<Event> eventNode) {
        eventNode.addListener(PlayerBlockInteractEvent.class, this::onInteract);
        eventNode.addListener(PlayerPacketEvent.class, this::onPacket);
        eventNode.addListener(EntityDismountEvent.class, this::onDismount);
        eventNode.addListener(PlayerDisconnectEvent.class, this::onDisconnect);

    }

    private void onPacket(PlayerPacketEvent event) {
        if (event.getPacket() instanceof ClientSteerVehiclePacket packet) {
            var ridingEntity = event.getPlayer().getVehicle();
            if (packet.flags() == 2 && ridingEntity != null) {
                var entityDismountEvent = new EntityDismountEvent(ridingEntity, event.getPlayer());
                EventDispatcher.call(entityDismountEvent);
            }
         }
    }

    private void onDisconnect(PlayerDisconnectEvent event) {
        removePlayer(event.getPlayer());
    }

    private void onDismount(EntityDismountEvent event) {
        if (event.rider() instanceof Player player && isSitting(player)) {
            removePlayer(player);
        }
    }

    private void onInteract(PlayerBlockInteractEvent event) {
        if (event.getBlock().namespace() == Block.SPRUCE_STAIRS.namespace()) {
            sitPlayer(event.getPlayer(), event.getBlockPosition());
        }
    }

    private void sitPlayer(Player player, Point sitLocation) {
        var instance = player.getInstance();
        if (instance == null) return;
        var playerLocation = player.getPosition();
        var arrow = new Entity(EntityType.ARROW);
        arrow.setInstance(instance, sitLocation.add(0.5, SIT_OFFSET, 0.5));
        arrow.setInvisible(true);
        arrow.setSilent(true);

        playerExitLocations.put(player.getUuid(), playerLocation);
        arrow.addPassenger(player);

        playerSitsMap.remove(player.getUuid());
        playerSitsMap.put(player.getUuid(), arrow);
    }

    private void removePlayer(Player player) {
        var arrow = playerSitsMap.getOrDefault(player.getUuid(), null);
        if (arrow != null) {
            arrow.removePassenger(player);
            arrow.remove();

            var exitLocation = playerExitLocations.remove(player.getUuid());
            if (exitLocation != null) {
                player.teleport(exitLocation);
            }
        }
    }

    private boolean isSitting(Player player) {
        return playerSitsMap.containsKey(player.getUuid());
    }

    public static SitFunction instance(EventNode<Event> eventNode) {
        return new SitFunction(eventNode);
    }

}

File: src/main/kotlin/net/onelitefeather/titan/function/SitFunction.kt
package net.onelitefeather.titan.function

import com.google.inject.Inject
import com.google.inject.name.Named
import net.minestom.server.coordinate.Point
import net.minestom.server.coordinate.Pos
import net.minestom.server.entity.Entity
import net.minestom.server.entity.EntityType
import net.minestom.server.entity.Player
import net.minestom.server.event.Event
import net.minestom.server.event.EventDispatcher
import net.minestom.server.event.EventListener
import net.minestom.server.event.EventNode
import net.minestom.server.event.player.PlayerBlockInteractEvent
import net.minestom.server.event.player.PlayerDisconnectEvent
import net.minestom.server.event.player.PlayerPacketEvent
import net.minestom.server.instance.block.Block
import net.minestom.server.network.packet.client.play.ClientSteerVehiclePacket
import net.onelitefeather.titan.event.EntityDismountEvent
import java.util.*

class SitFunction : TitanFunction {

    private val offsetY: Double = 0.25

    @Named("titanNode")
    @Inject
    lateinit var eventNode: EventNode<Event>

    private var playerSitsMap: MutableMap<UUID, Entity> = mutableMapOf()
    private var playerExitLocations: MutableMap<UUID, Pos> = mutableMapOf()

    private fun onDisconnect(playerDisconnectEvent: PlayerDisconnectEvent) {
        removePlayer(playerDisconnectEvent.player, true)
    }

    private fun onDismount(entityDismountEvent: EntityDismountEvent) {
        if (entityDismountEvent.rider is Player && isSitting(entityDismountEvent.rider)) {
            removePlayer(entityDismountEvent.rider, true)
        }
    }

    private fun onPacket(event: PlayerPacketEvent) {
        val packet = event.packet
        if (packet is ClientSteerVehiclePacket) {
            val ridingEntity = event.player.vehicle
            if (packet.flags == 2.toByte() && ridingEntity != null) {
                val entityDismount =
                    EntityDismountEvent(ridingEntity, event.player)
                EventDispatcher.call(entityDismount)
            }
        }
    }

    private fun onInteract(playerBlockInteractEvent: PlayerBlockInteractEvent) {
        if (playerBlockInteractEvent.block.namespace() == Block.SPRUCE_STAIRS.namespace()) {
            sitPlayer(
                playerBlockInteractEvent.player,
                playerBlockInteractEvent.blockPosition
            )
        }
    }

    private fun sitPlayer(player: Player, sitLocation: Point) {
        val instance = player.instance ?: return
        val playerLocation = player.position
        val arrow = Entity(EntityType.ARROW)
        arrow.setInstance(instance, sitLocation.add(0.5, offsetY, 0.5))
        arrow.isSilent = true
        arrow.isInvisible = true

        playerExitLocations[player.uuid] = playerLocation
        arrow.addPassenger(player)

        playerSitsMap[player.uuid]?.remove()
        playerSitsMap[player.uuid] = arrow
    }

    private fun isSitting(player: Player): Boolean {
        return playerSitsMap.containsKey(player.uuid)
    }

    private fun removePlayer(player: Player, remove: Boolean) {
        val arrow = playerSitsMap.remove(player.uuid)

        if (arrow != null) {
            arrow.removePassenger(player)
            if (remove) arrow.remove()

            val exitLocation = playerExitLocations.remove(player.uuid)
            if (exitLocation != null) {
                player.teleport(exitLocation)
            }
        }
    }

    override fun initialize() {
        eventNode.addListener(
            EventListener.of(
                PlayerBlockInteractEvent::class.java,
                this::onInteract
            )
        )
        eventNode.addListener(
            EventListener.of(
                PlayerPacketEvent::class.java,
                this::onPacket
            )
        )
        eventNode.addListener(
            EventListener.of(
                EntityDismountEvent::class.java,
                this::onDismount
            )
        )
        eventNode.addListener(
            EventListener.of(
                PlayerDisconnectEvent::class.java,
                this::onDisconnect
            )
        )
    }

    override fun terminate() {
    }
}